airbyte-integrations/connectors/source-slack/source_slack/source.py

```diff
@@ -0,0 +1,286 @@
+        the logic simple to reason about.
+
+        Good luck.
+        """
+
+        stream_state = stream_state or {}
+        channels_stream = Channels(authenticator=self.authenticator)
+        if "start_ts" in stream_state:
+            # Since new messages can be posted to threads continuously after the parent message has been posted, we get messages from the latest date
+            # found in the state minus 7 days to pick up any new messages in threads.
+            # If there is state always use lookback
+            messages_start_date = pendulum.from_timestamp(stream_state.get("start_ts")).subtract(**self.messages_lookback_window)
+        else:
+            # If there is no state i.e: this is the first sync then there is no use for lookback, just get messages from the default start date
+            messages_start_date = self._default_start_date
+        messages_stream = ChannelMessages(authenticator=self.authenticator, default_start_date=messages_start_date)
+
+        for message_chunk in messages_stream.batches(stream_state={"start_ts": messages_start_date.timestamp()}):
+            for channel in channels_stream.read_records(sync_mode=SyncMode.full_refresh):
+                message_chunk['channel'] = channel['id']
+                for message in messages_stream.read_records(sync_mode=SyncMode.full_refresh, batch=message_chunk):
+                    yield {"channel": channel['id'], "ts": message['ts']}
+
+
+class JoinChannelsStream(HttpStream):
+    """
+    This class is a special stream which joins channels because the Slack API only returns messages from channels this bot is in.
+    Its responses should only be logged for debugging reasons, not read as records.
+    """
+
+    url_base = "https://slack.com/api/"
+    http_method = "POST"
+
+    def parse_response(self, response: requests.Response, batch: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping]:
+        return [{"message": f"Successfully joined channel: {batch['channel_name']}"}]
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None  # No pagination
+
+    def path(self, **kwargs) -> str:
+        return "conversations.join"
+
+    def batches(self, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        channels_stream = Channels(authenticator=self.authenticator)
+        for channel in channels_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield {'channel': channel['id'], 'channel_name': channel['name']}
+
+    def request_body_json(self, batch: Mapping = None, **kwargs) -> Optional[Mapping]:
+        return {'channel': batch['channel']}
+
+
+class SourceSlack(AbstractSource):
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Optional[Any]]:
+        slack_client = WebClient(token=config["api_token"])
+        users = slack_client.users_list(limit=1).get("members", [])
```

Comment:
sanity check: does checking for members check we can do all of the above?

Reply:
it does not! #2765 exists to address exactly this point, but definitely a stretch for v1.